File: 2023-03-31-aistats-slopecd/scripts/pgd-vs-cd.jl
using Plots
using LaTeXStrings
using Parquet2
using DataFrames
using StatsPlots
using TableOperations

gr()

default(fontfamily="Computer Modern")

ds = Parquet2.Dataset("data/pgd-vs-cd-lasso.parquet")

obj_fun = function (x)
  return x .- minimum(x)
end

df = ds |>
     TableOperations.select(:solver_name, :data_name, :time, :objective_value) |>
     DataFrame

df = transform(df, :)

df.solver_name[df.solver_name.=="Python-PGD[use_acceleration=True]"] .= "PGD"
df.solver_name[df.solver_name.=="cd"] .= "CD"
df.data_name[df.data_name.=="libsvm[dataset=rcv1.binary]"] .= "RCV1"
df.data_name[df.data_name.=="libsvm[dataset=news20.binary]"] .= "news20"

suboptim(x) = x .- minimum(x) .+ 1e-10

gdf = groupby(df, :data_name)

pls = []

for df in gdf
  title = unique(df.data_name)[1]
  pl = @df df plot(
    :time,
    suboptim(:objective_value),
    group=:solver_name,
    yaxis=:log,
    title = title,
    xlabel="Time (s)",
    ylabel="Suboptimality"
  )
  push!(pls, pl)
end

# pls = [@df df plot(:time, suboptim(:objective_value), group=:solver_name, yaxis=:log, xlabel="Time (s)", title="Same", ylabel="Suboptimality") for df in gdf]
#
plot(
  pls...,
  layout=(2, 1),
  thickness_scaling=0.5,
  left_margin=Plots.cm,
  size=(150, 175),
  top_margin=Plots.mm,
  bottom_margin=0.5Plots.cm
)

path = "figures/cd-vs-pgd.pdf"

savefig(path)

# crop the figure
cmd = `pdfcrop $path $path`
run(cmd)

# @df df2 plot(
#   :time,
#   :objective_value_suboptim,
#   group = (:solver_name, :data_name),
#   yaxis=:log,
#   layout = 2,
#   legend=false,
#   ylabel="Suboptimality",
#   xlabel="Time (s)"
# )
#
#

File: 2023-03-31-aistats-slopecd/scripts/pgd-vs-cd.py
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from labellines import labelLines

from pythesis.utils import save_fig


def suboptim(x):
    return x - np.min(x) + 1e-10


data = pd.read_parquet("data/pgd-vs-cd-lasso.parquet")
data = data.filter(items=["data_name", "objective_value", "solver_name", "time"])

data["solver_name"].replace("Python-PGD[use_acceleration=True]", "PGD", inplace=True)
data["solver_name"].replace("cd", "CD", inplace=True)


news20 = data.query("data_name == 'libsvm[dataset=news20.binary]'").copy()
rcv1 = data.query("data_name == 'libsvm[dataset=rcv1.binary]'").copy()

news20["objective_value"] = suboptim(news20["objective_value"])
rcv1["objective_value"] = suboptim(rcv1["objective_value"])
# news20["objective_value"] = news20["objective_value"] - np.min(news20["objective_value"])

fig, axs = plt.subplots(
    1, 2, figsize=(3.5, 1.8), sharey=True, constrained_layout=True, edgecolor="white"
)

for key, grp in news20.groupby(["solver_name"]):
    axs[0].semilogy(grp["time"], grp["objective_value"], label=key)

for key, grp in rcv1.groupby(["solver_name"]):
    axs[1].semilogy(grp["time"], grp["objective_value"], label=key)

axs[0].set_title("news20")
axs[0].set_ylabel("Suboptimality")

axs[1].set_title("RCV1")

fig.supxlabel("Time (s)")

for i in range(2):
    labelLines(axs[i].get_lines())

path = "paper5-cd-vs-pgd.pdf"

save_fig(path)

plt.close("all")
